src/main/java/com/exasol/releasedroid/main/ReleaseDroid.java

```diff
@@ -51,21 +51,28 @@ public ReleaseDroid(final RepositoryGateway repositoryGateway, final UseCase val
     public void run(final UserInput userInput) {
         validateUserInput(userInput);
         final Repository repository = this.repositoryGateway.getRepository(userInput);
-
         final ReleasePlatforms platforms = ReleasePlatforms.from(userInput, repository);
         LOGGER.fine(() -> "Release Droid has received '" + userInput.getGoal() + "' request for the project '"
                 + userInput.getFullRepositoryName() + "'.");
         final List<Report> reports = new ArrayList<>();
-        final boolean isValidate = userInput.getGoal() != Goal.RELEASE;
-        final UseCase useCase = isValidate ? this.validateUseCase : this.releaseUseCase;
+
         final Optional<Path> releaseGuide = userInput.releaseGuide();
-        if (releaseGuide.isPresent() && isValidate) {
+        if (releaseGuide.isPresent()) {
             ReleaseGuide.from(repository).write(releaseGuide.get());
         }
-        reports.addAll(useCase.apply(repository, platforms));
+        // always apply use case "validate" first
+        reports.addAll(this.validateUseCase.apply(repository, platforms));
+        // apply use case "release" only if requested and validation showed no failures
+        if ((userInput.getGoal() == Goal.RELEASE) && !hasFailures(reports)) {
+            reports.addAll(this.releaseUseCase.apply(repository, platforms));
+        }
```

Comment:
Extract to a method with a speaking name instead on using inline comments.

Reply:
Good proposal, I will do so

Reply:
After looking at it more closely I am not convinced anymore. 
The extracted method would require four parameters (`userInput`, `reports`, `repository`, and `platforms`) and would need to modify `reports` which feels a little bit clumsy. What do you think?

Reply:
The code is readable enough as it is, so no real pressure to change. Keep it.